Samagra-Development/workflow
# Migrated to Turborepo (Monorepo Structure)

Signed-off-by: Ansh Sarkar <anshsark18@gmail.com>

Issues: #7 & #3

## Description

Migration of existing repository structure to a monorepo using **[Turborepo](https://turborepo.org/)** in order to improve scalability and Developer Experience (DX) by enabling division into workspaces, caching and faster builds.

### Changes

- [x] Added 3 workspaces in accordance to [recommended configuration of workspaces](https://turborepo.org/docs/getting-started/existing-monorepo#configure-workspaces).
- [x] Added ```turborepo``` for monorepo management and ```concurrent``` for executing parallel tasks in ```NodeJS```.
- [x] Added turborepo configuration in ```turbo.json``` for caching outputs of ```build``` and ```build-docs``` scripts.
- [x] Enabled concurrent execution of tasks. Improved build time.
- [x] Single command setup for packages ```enketo-core```, ```enketo-express``` and ```form-manager``` and sample application ```wrapper```.
- [x] Modified directory structure according to workspace configuration defined in ```package.json```.
- [x] Added ```npm run start``` for installing dependencies and starting all services.
- [x] Added ```npm run build``` for building necessary modules as well as documentations.

## How to test

- [x] Clone repository locally
```
git clone https://github.com/Samagra-Development/workflow.git
```
- [x] Switch to the repository directory
```
cd workflow
```
- [x] Start the project
```
npm run start
```

The above command will automatically install, set up and run all the packages and applications. ```Docker``` is required. After successful build you should be able to view the form at : http://localhost:3000/

**Note:** Initial build takes roughly 20 - 25 mins, due to ```enketo-core```. Subsequent builds take less than 1 min. The react frontend will show an error ```localhost refused to connect``` until ```enketo-core``` has finished building. Once the build is successful, refreshing the fronted will render the form. 

For building the modules and documentations you can run the ```build``` script as follow
```
npm run build
```
